Feature: Custom Columns

bschwarzent opened the pull request:
427023

bschwarzent committed:
TableUiPreferences: add withIgnoreTableEvents()

bschwarzent committed:
Table: check column type before calling type-specific methods

The pure existence of a property "backgroundEffect" does not mean that
_renderBackgroundEffect() can be called. Instead, check if the column
is instanceof NumberColumn.

bschwarzent committed:
Column: cleanup code

- Column: use protected modifier for method _ensureCell() to allow
  access in subclasses.
- ColumModel: improve jsdoc
- NumberColumn: always return a "nop" function for backgroundEffect,
  even if it is null. Otherwise, _cellStyle() will throw an error.

fschinkel committed:
Add Deferred class based on native ES6 promises

bschwarzent committed:
UiPreferences: stop scheduled task after spec

Avoid unexpected behavior caused by scheduled store tasks that are run
after the mock store has already been uninstalled.

bschwarzent committed:
TableCustomizer: allow changing column configurations programmatically

- Add methods to add/modify/remove column configurations
  programmatically, without user interaction.
- Existing methods were renamed ("custom columns" instead of just
  "columns").
- TableCustomizer can be specified in the table mode as an instance,
  a child model or an object type. (Not necessary for TableOrganizer
  because it has no mandatory constructor arguments.)
- TableCustomizer supports asynchronous creation of columns. Callers
  can wait for the process to be done with whenReady().
- Modernization: use native promises instead of jQuery promises.

427023

bschwarzent committed:
Table: fix types

bschwarzent committed:
Cleanup imports

bschwarzent committed:
JsPageHelper: preserve rows when child pages of a table page are removed

For a PageWithTable, the table is the data master. If child pages are
removed from the tree, the corresponding table rows must not be deleted
automatically. Instead, the rows are simply unlinked from the deleted
child pages.

Otherwise, reloading a hybrid table page would delete all rows when the
java child nodes are deleted on the ui server. This causes the current
row selection to be lost, which means it cannot be restored when the
table is finally populated again.

fschinkel commented on eclipse-scout-core/src/desktop/outline/pages/js/JsPageHelper.ts:
```diff
@@ -119,9 +119,11 @@ export class JsPageHelper implements JsPageHelperModel, ObjectWithType {
-    // remove corresponding rows, otherwise e.g. double-clicking a row will lead to errors as the page was deleted
-    const rows = pages.map(page => page.row).filter(Boolean);
-    this.page.detailTable.deleteRows(rows);
+    // remove or unlink corresponding rows, otherwise e.g. double-clicking a row will lead to errors as the page was deleted
```
Why does it say "remove or unlink"? The rows are not removed but only unlinked.

fschinkel commented on eclipse-scout-core/src/desktop/outline/pages/js/JsPageHelper.ts:
```diff
@@ -139,8 +141,11 @@ export class JsPageHelper implements JsPageHelperModel, ObjectWithType {
-    // remove all rows, otherwise e.g. double-clicking a row will lead to errors as the page was deleted
-    this.page.detailTable.deleteAllRows();
+    // remove or unlink all rows, otherwise e.g. double-clicking a row will lead to errors as the page was deleted
+    if (this.page instanceof PageWithTable) {
```
Why does it say "remove or unlink"? The rows are not removed but only unlinked.

fschinkel commented on eclipse-scout-core/test/desktop/outline/pages/js/JsPageHelperSpec.ts:
```diff
@@ -192,6 +193,8 @@ describe('JsPageHelper', () => {
       await myPageWithTable.ensureLoadChildren();
 
+      const row1 = myPageWithTable.detailTable.rows[0];
+      const row2 = myPageWithTable.detailTable.rows[1];
```
```suggestion
      const [row1, row2, row3, row4] = myPageWithTable.detailTable.rows;
```

fschinkel commented on eclipse-scout-core/test/desktop/outline/pages/js/JsPageHelperSpec.ts:
```diff
@@ -243,6 +307,52 @@ describe('JsPageHelper', () => {
       await myPageWithTable.ensureLoadChildren();
 
+      const row1 = myPageWithTable.detailTable.rows[0];
+      const row2 = myPageWithTable.detailTable.rows[1];
```
```suggestion
      const [row1, row2, row3, row4] = myPageWithTable.detailTable.rows;
```

cguglielmo commented on eclipse-scout-core/src/table/Table.ts:
```diff
@@ -2911,7 +2911,10 @@ export class Table extends Widget implements TableModel, Filterable<TableRow> {
   protected _updateBackgroundEffect() {
-    this.columns.forEach((column: NumberColumn) => {
+    this.columns.forEach(column => {
+      if (!(column instanceof NumberColumn)) {
```
Are there other columns with a property backgroundEffect?

cguglielmo commented on eclipse-scout-core/src/table/columns/NumberColumn.ts:
```diff
@@ -223,10 +223,10 @@ export class NumberColumn extends Column<number> implements NumberColumnModel {
     }
 
-    if (effect !== null) {
+    if (effect) {
```
Should be ok, but are there other effects than colorGradient1, colorGradient2 and barChart that are not handled here?

cguglielmo commented on eclipse-scout-core/src/table/columns/ColumnModel.ts:
```diff
@@ -214,6 +214,10 @@ export interface ColumnModel<TValue = string> extends ObjectModelWithUuid<Column
 
+  /**
+   * Specifies whether this column can be used as the "node column" of a hierarchical table, i.e. the column showing
+   * the indentation and the table row controls. Default is true.
```
Please write "Default is true" at the end separated by an empty line to be consistent with the other methods.

cguglielmo commented on eclipse-scout-core/src/table/customizer/TableCustomizer.ts:
```diff
@@ -30,26 +38,169 @@ export abstract class TableCustomizer implements TableCustomizerModel, ObjectWit
+
+export interface TableCustomizerCreateColumnsOptions {
+  /**
+   * default is false
```
Please capitalize first letter and add a dot at the end.

cguglielmo commented on eclipse-scout-core/src/table/organizer/TableOrganizer.ts:
```diff
@@ -255,15 +259,15 @@ export class TableOrganizer implements ObjectWithType {
     });
     form.open();
-    return form.whenSave().then(() => {
+    await form.whenSave().then(() => {
```
Does this still return the whenSave promise?

bschwarzent committed:
TableUiPreferences: ignore non-displayable columns

Columns with displayable=false are never visible to the user. Therefore,
it is unnecessary to store any information about these columns in the
ui preferences. They are neither relevant nor can they be changed. To
make things more deterministic, non-displayable columns are always
placed at the beginning of the 'columns' array when preferences are
applied.

The only exception are the the initial ui preferences of a table. We
deliberately include non-displayable columns here to be able to restore
the initial state exactly when the user invokes the 'restore to factory
settings' menu. In the case of bookmark, this also includes bookmarked
user filters.

Columns that don't appear in the preferences by default appear at the
end of the table. This can happen e.g. if a new column was added to the
application after a user already changed the table's preferences. The
sort algorithm was improved to allow new columns to be placed at the
front of the table. To applies to all columns *without* preferences
that appear in the original column list before the first column *with*
preferences. All other new columns are still placed at the end.

bschwarzent committed:
Add ObjectRegistry

The object registry represents an ordered collection of instances of
the specified object type. These are created automatically on first
access and are stored in the registry indefinitely or until explicitly
unregistered.

The object registry can be used as an extension point for other
application modules. It allows callers to apply the "inversion of
control" principle, similar to BEANS.all() in Java.
